File: src/main/java/cc/mirukuneko/realtrainmodrenewed/RealTrainModRenewedItems.java
package cc.mirukuneko.realtrainmodrenewed;

import cc.mirukuneko.realtrainmodrenewed.item.CarItem;
import cc.mirukuneko.realtrainmodrenewed.item.CrowbarItem;
import cc.mirukuneko.realtrainmodrenewed.item.IcCardItem;
import cc.mirukuneko.realtrainmodrenewed.item.MarkerItem;
import cc.mirukuneko.realtrainmodrenewed.item.RailItem;
import cc.mirukuneko.realtrainmodrenewed.item.TrainItem;
import cc.mirukuneko.realtrainmodrenewed.item.InstalledObjectItem;
import cc.mirukuneko.realtrainmodrenewed.item.TrainVehicleItem;
import cc.mirukuneko.realtrainmodrenewed.item.WireItem;
import cc.mirukuneko.realtrainmodrenewed.item.WrenchItem;
import cc.mirukuneko.realtrainmodrenewed.installedobject.InstalledObjectCategory;
import net.neoforged.neoforge.registries.DeferredItem;
import net.neoforged.neoforge.registries.DeferredRegister;

public class RealTrainModRenewedItems {

    public static final DeferredRegister.Items ITEMS = DeferredRegister.createItems(RealTrainModRenewed.MODID);

    public static final DeferredItem<InstalledObjectItem> CROSSING_GATE_ITEM = ITEMS.registerItem(
        "crossing_gate", props -> new InstalledObjectItem(InstalledObjectCategory.CROSSING, props)
    );
    // 以下のアイテムはユーザー要望により削除:
    //   受信機(signal_receiver) / 受信機シグナル値(signal_value_receiver) / 電車検知ブロック(train_detector)
    //   状態ブロック(signal_state) / スクリプトブロック(script_block) / 通信機(signal_communicator)
    // 道床(ballast)アイテムも廃止済み。ブロック自体は残るがアイテム(入手手段)は登録しない。
    public static final DeferredItem<MarkerItem> MARKER_ITEM = ITEMS.registerItem(
        "marker", props -> new MarkerItem(RealTrainModRenewedBlocks.MARKER.get(), false, props)
    );
    public static final DeferredItem<MarkerItem> MARKER_DIAGONAL_ITEM = ITEMS.registerItem(
        "marker_diagonal", props -> new MarkerItem(RealTrainModRenewedBlocks.MARKER.get(), true, props)
    );
    public static final DeferredItem<MarkerItem> MARKER_SWITCH_ITEM = ITEMS.registerItem(
        "marker_switch", props -> new MarkerItem(RealTrainModRenewedBlocks.MARKER_SWITCH.get(), false, props)
    );
    public static final DeferredItem<MarkerItem> MARKER_SWITCH_DIAGONAL_ITEM = ITEMS.registerItem(
        "marker_switch_diagonal", props -> new MarkerItem(RealTrainModRenewedBlocks.MARKER_SWITCH.get(), true, props)
    );
    public static final DeferredItem<RailItem> RAIL_ITEM = ITEMS.registerItem(
        "rail", RailItem::new
    );
    public static final DeferredItem<TrainItem> TRAIN_ITEM = ITEMS.registerItem(
        "train", props -> new TrainItem(TrainItem.Category.ELECTRIC, props)
    );
    // 試験用車両(test_train)はユーザー要望により削除。
    public static final DeferredItem<TrainVehicleItem> TRAIN_VEHICLE_ITEM = ITEMS.registerItem(
        "train_vehicle", props -> new TrainVehicleItem(props.stacksTo(1))
    );
    public static final DeferredItem<CarItem> CAR_ITEM = ITEMS.registerItem(
        "car", props -> new CarItem(props.stacksTo(1))
    );
    public static final DeferredItem<IcCardItem> IC_CARD_ITEM = ITEMS.registerItem(
        "ic_card", props -> new IcCardItem(props.stacksTo(1))
    );
    public static final DeferredItem<CrowbarItem> CROWBAR_ITEM = ITEMS.registerItem(
        "crowbar", props -> new CrowbarItem(props.stacksTo(1))
    );
    public static final DeferredItem<WrenchItem> WRENCH_ITEM = ITEMS.registerItem(
        "wrench", props -> new WrenchItem(props.stacksTo(1))
    );
    public static final DeferredItem<WireItem> WIRE_ITEM = ITEMS.registerItem(
        "wire", WireItem::new
    );
    // 照明(light): 本家RTM の照明アイテム。外部パックのモデルを使用し、レッドストーン
    // 信号を受けると点灯する(InstalledObjectBlock 側で発光処理)。
    public static final DeferredItem<InstalledObjectItem> LIGHT_ITEM = ITEMS.registerItem(
        "light", props -> new InstalledObjectItem(InstalledObjectCategory.LIGHT, props)
    );
    public static final DeferredItem<InstalledObjectItem> INSULATOR_ITEM = ITEMS.registerItem(
        "insulator", props -> new InstalledObjectItem(InstalledObjectCategory.INSULATOR, props)
    );
    public static final DeferredItem<InstalledObjectItem> SIGNAL_ITEM = ITEMS.registerItem(
        "signal", props -> new InstalledObjectItem(InstalledObjectCategory.SIGNAL, props)
    );
    public static final DeferredItem<InstalledObjectItem> OVERHEAD_LINE_POLE_ITEM = ITEMS.registerItem(
        "overhead_line_pole", props -> new InstalledObjectItem(InstalledObjectCategory.OVERHEAD_LINE_POLE, props)
    );
    public static final DeferredItem<InstalledObjectItem> TICKET_GATE_ITEM = ITEMS.registerItem(
        "ticket_gate", props -> new InstalledObjectItem(InstalledObjectCategory.TICKET_GATE, props)
    );
    public static final DeferredItem<InstalledObjectItem> SPEAKER_ITEM = ITEMS.registerItem(
        "speaker", props -> new InstalledObjectItem(InstalledObjectCategory.SPEAKER, props)
    );
}

File: src/main/kotlin/cc/mirukuneko/realtrainmodrenewed/RealTrainModRenewedItems.kt
// SPDX-License-Identifier: LGPL-3.0-or-later
// Copyright © 2026 mirukuneko and RealTrainModRenewed contributors
package cc.mirukuneko.realtrainmodrenewed

import cc.mirukuneko.realtrainmodrenewed.item.CarItem
import cc.mirukuneko.realtrainmodrenewed.item.CrowbarItem
import cc.mirukuneko.realtrainmodrenewed.item.IcCardItem
import cc.mirukuneko.realtrainmodrenewed.item.MarkerItem
import cc.mirukuneko.realtrainmodrenewed.item.RailItem
import cc.mirukuneko.realtrainmodrenewed.item.TrainItem
import cc.mirukuneko.realtrainmodrenewed.item.InstalledObjectItem
import cc.mirukuneko.realtrainmodrenewed.item.TrainVehicleItem
import cc.mirukuneko.realtrainmodrenewed.item.WireItem
import cc.mirukuneko.realtrainmodrenewed.item.WrenchItem
import cc.mirukuneko.realtrainmodrenewed.item.SignalConverterItem
import cc.mirukuneko.realtrainmodrenewed.electric.SignalConverterType
import cc.mirukuneko.realtrainmodrenewed.installedobject.InstalledObjectCategory
import net.minecraft.world.item.Item
import net.minecraft.world.item.BlockItem
import net.neoforged.neoforge.registries.DeferredItem
import net.neoforged.neoforge.registries.DeferredRegister
import java.util.function.Function
import java.util.function.Supplier

object RealTrainModRenewedItems {
    @JvmField
    val ITEMS: DeferredRegister.Items = DeferredRegister.createItems(RealTrainModRenewed.MODID)

    @JvmField val CROSSING_GATE_ITEM: DeferredItem<InstalledObjectItem> = ITEMS.registerItem("crossing_gate", Function { props: Item.Properties -> InstalledObjectItem(InstalledObjectCategory.CROSSING, props) })
    @JvmField val MARKER_ITEM: DeferredItem<MarkerItem> = ITEMS.registerItem("marker", Function { props: Item.Properties -> MarkerItem(RealTrainModRenewedBlocks.MARKER.get(), false, props) })
    @JvmField val MARKER_DIAGONAL_ITEM: DeferredItem<MarkerItem> = ITEMS.registerItem("marker_diagonal", Function { props: Item.Properties -> MarkerItem(RealTrainModRenewedBlocks.MARKER.get(), true, props) })
    @JvmField val MARKER_SWITCH_ITEM: DeferredItem<MarkerItem> = ITEMS.registerItem("marker_switch", Function { props: Item.Properties -> MarkerItem(RealTrainModRenewedBlocks.MARKER_SWITCH.get(), false, props) })
    @JvmField val MARKER_SWITCH_DIAGONAL_ITEM: DeferredItem<MarkerItem> = ITEMS.registerItem("marker_switch_diagonal", Function { props: Item.Properties -> MarkerItem(RealTrainModRenewedBlocks.MARKER_SWITCH.get(), true, props) })
    @JvmField val RAIL_ITEM: DeferredItem<RailItem> = ITEMS.registerItem("rail", Function { props: Item.Properties -> RailItem(props) })
    @JvmField val TRAIN_ITEM: DeferredItem<TrainItem> = ITEMS.registerItem("train", Function { props: Item.Properties -> TrainItem(TrainItem.Category.ELECTRIC, props) })
    @JvmField val TRAIN_VEHICLE_ITEM: DeferredItem<TrainVehicleItem> = ITEMS.registerItem("train_vehicle", Function { props: Item.Properties -> TrainVehicleItem(props) })
    @JvmField val CAR_ITEM: DeferredItem<CarItem> = ITEMS.registerItem("car", Function { props: Item.Properties -> CarItem(props) })
    @JvmField val IC_CARD_ITEM: DeferredItem<IcCardItem> = ITEMS.registerItem("ic_card", Function { props: Item.Properties -> IcCardItem(props) })
    @JvmField val CROWBAR_ITEM: DeferredItem<CrowbarItem> = ITEMS.registerItem("crowbar", Function { props: Item.Properties -> CrowbarItem(props) })
    @JvmField val WRENCH_ITEM: DeferredItem<WrenchItem> = ITEMS.registerItem("wrench", Function { props: Item.Properties -> WrenchItem(props) })
    @JvmField val WIRE_ITEM: DeferredItem<WireItem> = ITEMS.registerItem("wire", Function { props: Item.Properties -> WireItem(props) })
    @JvmField val SIGNAL_CONVERTER_ITEM: DeferredItem<SignalConverterItem> = ITEMS.registerItem("signal_converter", Function { props: Item.Properties -> SignalConverterItem(RealTrainModRenewedBlocks.SIGNAL_CONVERTER.get(), SignalConverterType.RS_INPUT, props) })
    @JvmField val SIGNAL_CONVERTER_RS_ITEM: DeferredItem<SignalConverterItem> = ITEMS.registerItem("signal_converter_rs", Function { props: Item.Properties -> SignalConverterItem(RealTrainModRenewedBlocks.SIGNAL_CONVERTER.get(), SignalConverterType.RS_OUTPUT, props) })
    @JvmField val SIGNAL_CONVERTER_INCREMENT_ITEM: DeferredItem<SignalConverterItem> = ITEMS.registerItem("signal_converter_increment", Function { props: Item.Properties -> SignalConverterItem(RealTrainModRenewedBlocks.SIGNAL_CONVERTER.get(), SignalConverterType.INCREMENT, props) })
    @JvmField val SIGNAL_CONVERTER_DECREMENT_ITEM: DeferredItem<SignalConverterItem> = ITEMS.registerItem("signal_converter_decrement", Function { props: Item.Properties -> SignalConverterItem(RealTrainModRenewedBlocks.SIGNAL_CONVERTER.get(), SignalConverterType.DECREMENT, props) })
    @JvmField val SIGNAL_CONVERTER_WIRELESS_ITEM: DeferredItem<SignalConverterItem> = ITEMS.registerItem("signal_converter_wireless", Function { props: Item.Properties -> SignalConverterItem(RealTrainModRenewedBlocks.SIGNAL_CONVERTER.get(), SignalConverterType.WIRELESS, props) })
    @JvmField val SIGNAL_RECEIVER_ITEM: DeferredItem<BlockItem> = ITEMS.registerItem("signal_receiver", Function { props: Item.Properties -> BlockItem(RealTrainModRenewedBlocks.SIGNAL_RECEIVER.get(), props) })
    @JvmField val SIGNAL_CHANGER_ITEM: DeferredItem<BlockItem> = ITEMS.registerItem("signal_changer", Function { props: Item.Properties -> BlockItem(RealTrainModRenewedBlocks.SIGNAL_CHANGER.get(), props) })
    @JvmField val SIGNAL_VALUE_RECEIVER_ITEM: DeferredItem<BlockItem> = ITEMS.registerItem("signal_value_receiver", Function { props: Item.Properties -> BlockItem(RealTrainModRenewedBlocks.SIGNAL_VALUE_RECEIVER.get(), props) })
    @JvmField val TRAIN_DETECTOR_ITEM: DeferredItem<BlockItem> = ITEMS.registerItem("train_detector", Function { props: Item.Properties -> BlockItem(RealTrainModRenewedBlocks.TRAIN_DETECTOR.get(), props) })
    @JvmField val SIGNAL_STATE_ITEM: DeferredItem<BlockItem> = ITEMS.registerItem("signal_state", Function { props: Item.Properties -> BlockItem(RealTrainModRenewedBlocks.SIGNAL_STATE.get(), props) })
    @JvmField val LIGHT_ITEM: DeferredItem<InstalledObjectItem> = ITEMS.registerItem("light", Function { props: Item.Properties -> InstalledObjectItem(InstalledObjectCategory.LIGHT, props) })
    @JvmField val INSULATOR_ITEM: DeferredItem<InstalledObjectItem> = ITEMS.registerItem("insulator", Function { props: Item.Properties -> InstalledObjectItem(InstalledObjectCategory.INSULATOR, props) })
    @JvmField val SIGNAL_ITEM: DeferredItem<InstalledObjectItem> = ITEMS.registerItem("signal", Function { props: Item.Properties -> InstalledObjectItem(InstalledObjectCategory.SIGNAL, props) })
    @JvmField val OVERHEAD_LINE_POLE_ITEM: DeferredItem<InstalledObjectItem> = ITEMS.registerItem("overhead_line_pole", Function { props: Item.Properties -> InstalledObjectItem(InstalledObjectCategory.OVERHEAD_LINE_POLE, props) })
    @JvmField val TICKET_GATE_ITEM: DeferredItem<InstalledObjectItem> = ITEMS.registerItem("ticket_gate", Function { props: Item.Properties -> InstalledObjectItem(InstalledObjectCategory.TICKET_GATE, props) })
    @JvmField val SPEAKER_ITEM: DeferredItem<InstalledObjectItem> = ITEMS.registerItem("speaker", Function { props: Item.Properties -> InstalledObjectItem(InstalledObjectCategory.SPEAKER, props) })
    @JvmField val SCRIPT_BLOCK_ITEM: DeferredItem<BlockItem> = ITEMS.registerItem("script_block", Function { props: Item.Properties -> BlockItem(RealTrainModRenewedBlocks.SCRIPT_BLOCK.get(), props) })
}
